src/Base/Rotation.cpp

```diff
@@ -220,32 +220,48 @@ void Rotation::setValue(const double q[4])
 
 void Rotation::setValue(const Matrix4D & m)
 {
-    double trace = (m[0][0] + m[1][1] + m[2][2]);
+    Matrix4D mc(m);
+    mc.setCol(3, Vector3d(0.0, 0.0, 0.0));
+    auto type = mc.hasScale();
+    if (type == Base::ScaleType::Uniform) {
+        // Scaled matrices results in a erroneus quaternion
+        // Removing scale make it work if scale is uniform
+        // or if there is no rotation
+        double d = mc.determinant3();
+        double s = 1.0/copysign(mc.getCol(0).Length(), d);
+        mc.scale(s, s, s);
+    } else if (type != Base::ScaleType::NoScaling) {
+        // Any other scale type are ambiguous about rotation

```

Comment:
I'm not skilled enough to say if this is completely true or not. :)
But this is good way to clearly drop cases that we don't correctly handle.
This is necessary though IMO to throw an exception in this case so caller is informed something went wrong.

Reply:
I try to filter out the cases when there is a uniform scale but no rotation and throw an exception for more complex cases.

Reply:
Obviously you tried to get the inverse scaling factors of a non-uniform scaling but then run into problems. Anyway there must be a way to also handle non-uniform scaling but it must be distinguished whether the scaling matrix has been multiplied from the left or the right side.

Here is an example in Python:
```
from math import copysign

# this is a pure rotation matrix
mat = App.Matrix()
mat.rotateX(1.234)
mat.rotateY(-4.321)
mat.rotateZ(1.423)

# this is a pure scaling matrix. The expected values for the inverse scaling are 1, 1/2 and 1/3
scale = App.Matrix()
scale.scale(1,2,3)

# multiply from left
mat1 = scale * mat
mat1.hasScale()

# and right
mat2 = mat * scale
mat2.hasScale()

# Scale type=NonUniformLeft:
# to get the inverse scaling factor one must use the rows of the matrix
# according to the changes of the first commit this would be
mat3 = App.Matrix(mat1)
dx = copysign(mat3.row(0).Length, mat3.A11)
dy = copysign(mat3.row(1).Length, mat3.A22)
dz = copysign(mat3.row(2).Length, mat3.A33)
mat3.scale(1/dx, 1/dy, 1/dz)
mat3.hasScale()

# mat3 now is indeed a pure rotation matrix which you can confirm with
mat3.col(0) * mat3.col(1) # = 0
mat3.col(0) * mat3.col(2) # = 0
mat3.col(1) * mat3.col(2) # = 0
mat3.determinant() # = 1

# But unfortunately it's a completely different rotation. This happens because dx=-1 and dz=-3 instead of the expected values of 1 and 3
If a scale factor is negative then it means it also does a mirroring and the question is whether this should be considered.

# Scale type=NonUniformRight:
mat4 = App.Matrix(mat2)
dx = copysign(mat4.col(0).Length, mat4.A11)
dy = copysign(mat4.col(1).Length, mat4.A22)
dz = copysign(mat4.col(2).Length, mat4.A33)
mat4.scale(1/dx, 1/dy, 1/dz)
mat4.hasScale()

```

Reply:
Yes. Obviously :)
Ok. I see. So it is possible to solve the rotation as long as the scale type is not `Other´. That doesn't seem too hard to do. I'll try to night.

Reply:
@wwmayer I'm thinking though before I start. Is it really relevant to extract the rotation in those cases, in particular the NonUniformLeft case? Maybe better to create a resolve method in Matrix class to extract scale, pos, rotation, etc?

Reply:
> Is it really relevant to extract the rotation in those cases, in particular the NonUniformLeft case?

If it's doable why not?

> Maybe better to create a resolve method in Matrix class to extract scale, pos, rotation, etc?

I agree. Doing it in the Rotation class is the very wrong place.

Your PR does it solve the issues you wanted to handle? If yes I can do a review of the current solution and once merged I can do the refactoring, extend the analysis to non-uniform scaling and write the C++ unit tests.

Reply:
> If it's doable why not?

Afrer some more thought I agree

> Your PR does it solve the issues you wanted to handle? If yes I can do a review of the current solution and once merged I can do the refactoring, extend the analysis to non-uniform scaling and write the C++ unit tests.

If there is no rush I would really like to try the extended analysis again, but yes it solves the particular case from the forum thread.

Reply:
OK, I can wait :)